SignalMessaging/profiles/OWSProfileManager.m

```diff
@@ -787,87 +566,70 @@ - (void)addUsersToProfileWhitelist:(NSArray<NSString *> *)recipientIds
-                if (![self isUserInProfileWhitelist:recipientId]) {
-                    [newRecipientIds addObject:recipientId];
-                }
-            }
-
-            if (newRecipientIds.count < 1) {
-                return;
-            }
-
-            for (NSString *recipientId in recipientIds) {
-                self.userProfileWhitelistCache[recipientId] = @(YES);
+    NSMutableSet<NSString *> *newRecipientIds = [NSMutableSet new];
+    [self.dbConnection asyncReadWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
+        for (NSString *recipientId in recipientIds) {
+            NSNumber *_Nullable oldValue =
+                [transaction objectForKey:recipientId inCollection:kOWSProfileManager_UserWhitelistCollection];
+            if (oldValue && oldValue.boolValue) {
+                continue;
             }
+            [transaction setObject:@(YES) forKey:recipientId inCollection:kOWSProfileManager_UserWhitelistCollection];
+            [newRecipientIds addObject:recipientId];
         }
-
-        [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
-            for (NSString *recipientId in recipientIds) {
-                [transaction setObject:@(YES)
-                                forKey:recipientId
-                          inCollection:kOWSProfileManager_UserWhitelistCollection];
+    }
+        completionBlock:^{
+            for (NSString *recipientId in newRecipientIds) {
+                [[NSNotificationCenter defaultCenter]
+                    postNotificationNameAsync:kNSNotificationName_ProfileWhitelistDidChange
+                                       object:nil
+                                     userInfo:@{
+                                         kNSNotificationKey_ProfileRecipientId : recipientId,
+                                     }];
             }
         }];
-
-        for (NSString *recipientId in newRecipientIds) {
-            [[NSNotificationCenter defaultCenter]
-                postNotificationNameAsync:kNSNotificationName_ProfileWhitelistDidChange
-                                   object:nil
-                                 userInfo:@{
-                                     kNSNotificationKey_ProfileRecipientId : recipientId,
-                                 }];
-        }
-    });
 }
 
 - (BOOL)isUserInProfileWhitelist:(NSString *)recipientId
 {
     OWSAssert(recipientId.length > 0);
 
-    @synchronized(self)
-    {
-        NSNumber *_Nullable value = self.userProfileWhitelistCache[recipientId];
-        if (value) {
-            return [value boolValue];
-        }
-
-        value =
-            @([self.dbConnection hasObjectForKey:recipientId inCollection:kOWSProfileManager_UserWhitelistCollection]);
-        self.userProfileWhitelistCache[recipientId] = value;
-        return [value boolValue];
-    }
+    __block BOOL result = NO;
+    [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
```

Comment:
Getting rid of the caches simplifies things. I guess it was a premature optimization?

Reply:
I think we want `readWithBlock:` here.

Reply:
Done.